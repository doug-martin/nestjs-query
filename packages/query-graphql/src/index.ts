export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
