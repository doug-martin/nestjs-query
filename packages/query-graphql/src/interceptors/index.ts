export * from './hook.interceptor';
export * from './authorizer.interceptor';
