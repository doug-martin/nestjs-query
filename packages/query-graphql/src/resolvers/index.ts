export { CRUDResolverOpts, CRUDResolver } from './crud.resolver';
export { CreateResolver, CreateResolverOpts } from './create.resolver';
export { ReadResolver, ReadResolverOpts } from './read.resolver';
export { UpdateResolver, UpdateResolverOpts } from './update.resolver';
export { DeleteResolver, DeleteResolverOpts } from './delete.resolver';
export { ReferenceResolver, ReferenceResolverOpts } from './reference.resolver';
export { Relatable, ResolverRelation, ResolverRelationReference } from './relations';
export { ResolverOpts } from './resolver.interface';
export { FederationResolver } from './federation';
