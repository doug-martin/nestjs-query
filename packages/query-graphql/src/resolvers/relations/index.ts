export { Relatable } from './relations.resolver';
export { ReadRelationsResolver } from './read-relations.resolver';
export { UpdateRelationsResolver } from './update-relations.resolver';
export { RemoveRelationsResolver } from './remove-relations.resolver';
export { ReferencesRelationsResolver } from './references-relation.resolver';
export { AggregateRelationsResolver } from './aggregate-relations.resolver';
export {
  RelationTypeMap,
  ReferencesOpts,
  ResolverRelationReference,
  RelationsOpts,
  ResolverRelation,
} from './relations.interface';
