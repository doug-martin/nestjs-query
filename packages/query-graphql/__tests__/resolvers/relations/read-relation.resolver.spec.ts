import { Query, Resolver } from '@nestjs/graphql';
import { deepEqual, objectContaining, when } from 'ts-mockito';
import { CursorQueryArgsType, NonePagingQueryArgsType, OffsetQueryArgsType, PagingStrategies } from '../../../src';
import { ReadRelationsResolver, RelationsOpts } from '../../../src/resolvers/relations';
import {
  generateSchema,
  createResolverFromNest,
  TestResolverDTO,
  TestService,
  TestRelationDTO,
} from '../../__fixtures__';

describe('ReadRelationsResolver', () => {
  const expectResolverSDL = async (opts?: RelationsOpts) => {
    @Resolver(() => TestResolverDTO)
    class TestSDLResolver extends ReadRelationsResolver(TestResolverDTO, opts ?? {}) {
      @Query(() => TestResolverDTO)
      test(): TestResolverDTO {
        return { id: '1', stringField: 'foo' };
      }
    }

    const schema = await generateSchema([TestSDLResolver]);
    expect(schema).toMatchSnapshot();
  };

  it('should not add read methods if one and many are empty', () => expectResolverSDL());
  describe('one', () => {
    @Resolver(() => TestResolverDTO)
    class TestResolver extends ReadRelationsResolver(TestResolverDTO, {
      one: { relation: { DTO: TestRelationDTO }, custom: { DTO: TestRelationDTO, relationName: 'other' } },
    }) {
      constructor(service: TestService) {
        super(service);
      }
    }
    it('should use the object type name', () => expectResolverSDL({ one: { relation: { DTO: TestRelationDTO } } }));

    it('should use the dtoName if provided', () =>
      expectResolverSDL({ one: { relation: { DTO: TestRelationDTO, dtoName: 'Test' } } }));

    it('should set the field to nullable if set to true', () =>
      expectResolverSDL({ one: { relation: { DTO: TestRelationDTO, nullable: true } } }));

    it('should not add read one methods if disableRead is true', () =>
      expectResolverSDL({ one: { relation: { DTO: TestRelationDTO, disableRead: true } } }));

    it('should call the service findRelation with the provided dto', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver);
      const dto: TestResolverDTO = {
        id: 'id-1',
        stringField: 'foo',
      };
      const output: TestRelationDTO = {
        id: 'id-2',
        testResolverId: dto.id,
      };
      when(
        mockService.findRelation(TestRelationDTO, 'relation', deepEqual([dto]), deepEqual({ filter: undefined })),
      ).thenResolve(new Map([[dto, output]]));
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.findRelation(dto, {});
      return expect(result).toEqual(output);
    });

    it('should call the service findRelation with the provided dto and correct relation name', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver);
      const dto: TestResolverDTO = {
        id: 'id-1',
        stringField: 'foo',
      };
      const output: TestRelationDTO = {
        id: 'id-2',
        testResolverId: dto.id,
      };
      when(
        mockService.findRelation(TestRelationDTO, 'other', deepEqual([dto]), deepEqual({ filter: undefined })),
      ).thenResolve(new Map([[dto, output]]));
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.findCustom(dto, {});
      return expect(result).toEqual(output);
    });
  });

  describe('many', () => {
    it('should use the object type name', () => expectResolverSDL({ many: { relations: { DTO: TestRelationDTO } } }));

    it('should use the dtoName if provided', () =>
      expectResolverSDL({ many: { relations: { DTO: TestRelationDTO, dtoName: 'Test' } } }));

    it('should set the field to nullable if set to true', () =>
      expectResolverSDL({ many: { relations: { DTO: TestRelationDTO, nullable: true } } }));

    it('should use an offset connection if pagingStrategy is offset', () =>
      expectResolverSDL({
        many: { relations: { DTO: TestRelationDTO, nullable: true, pagingStrategy: PagingStrategies.OFFSET } },
      }));

    it('should not add read methods if disableRead is true', () =>
      expectResolverSDL({ many: { relations: { DTO: TestRelationDTO, disableRead: true } } }));

    describe('many connection query', () => {
      @Resolver(() => TestResolverDTO)
      class TestResolver extends ReadRelationsResolver(TestResolverDTO, {
        one: { relation: { DTO: TestRelationDTO }, custom: { DTO: TestRelationDTO, relationName: 'other' } },
        many: { relations: { DTO: TestRelationDTO }, customs: { DTO: TestRelationDTO, relationName: 'others' } },
      }) {
        constructor(service: TestService) {
          super(service);
        }
      }

      it('should call the service queryRelations with the provided dto', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: CursorQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
          paging: { first: 1 },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(
            TestRelationDTO,
            'relations',
            deepEqual([dto]),
            objectContaining({ ...query, paging: { limit: 2, offset: 0 } }),
          ),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryRelations(dto, query, {});
        return expect(result).toEqual({
          edges: [
            {
              cursor: 'YXJyYXljb25uZWN0aW9uOjA=',
              node: {
                id: output[0].id,
                testResolverId: dto.id,
              },
            },
          ],
          pageInfo: {
            endCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
            hasNextPage: false,
            hasPreviousPage: false,
            startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
          },
          totalCountFn: expect.any(Function),
        });
      });

      it('should call the service countRelations with the provided dto', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: CursorQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
          paging: { first: 1 },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(
            TestRelationDTO,
            'relations',
            deepEqual([dto]),
            objectContaining({ ...query, paging: { limit: 2, offset: 0 } }),
          ),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryRelations(dto, query, {});
        when(
          mockService.countRelations(TestRelationDTO, 'relations', deepEqual([dto]), objectContaining(query.filter!)),
        ).thenResolve(new Map([[dto, 10]]));
        return expect(result.totalCount).resolves.toBe(10);
      });

      it('should call the service findRelation with the provided dto and correct relation name', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: CursorQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
          paging: { first: 1 },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(
            TestRelationDTO,
            'others',
            deepEqual([dto]),
            objectContaining({ ...query, paging: { limit: 2, offset: 0 } }),
          ),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryCustoms(dto, query, {});
        return expect(result).toEqual({
          edges: [
            {
              cursor: 'YXJyYXljb25uZWN0aW9uOjA=',
              node: {
                id: output[0].id,
                testResolverId: dto.id,
              },
            },
          ],
          pageInfo: {
            endCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
            hasNextPage: false,
            hasPreviousPage: false,
            startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
          },
          totalCountFn: expect.any(Function),
        });
      });
    });

    describe('many limit offset query', () => {
      @Resolver(() => TestResolverDTO)
      class TestResolver extends ReadRelationsResolver(TestResolverDTO, {
        one: { relation: { DTO: TestRelationDTO }, custom: { DTO: TestRelationDTO, relationName: 'other' } },
        many: {
          relations: { DTO: TestRelationDTO, pagingStrategy: PagingStrategies.OFFSET },
          customs: { DTO: TestRelationDTO, relationName: 'others', pagingStrategy: PagingStrategies.OFFSET },
        },
      }) {
        constructor(service: TestService) {
          super(service);
        }
      }

      it('should call the service queryRelations with the provided dto', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: OffsetQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
          paging: { limit: 1 },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(
            TestRelationDTO,
            'relations',
            deepEqual([dto]),
            objectContaining({ ...query, paging: { limit: 2 } }),
          ),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryRelations(dto, query, {});
        return expect(result).toEqual({
          nodes: output,
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false,
          },
          totalCountFn: expect.any(Function),
        });
      });

      it('should call the service findRelation with the provided dto and correct relation name', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: OffsetQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
          paging: { limit: 1 },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(
            TestRelationDTO,
            'others',
            deepEqual([dto]),
            objectContaining({ ...query, paging: { limit: 2 } }),
          ),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryCustoms(dto, query, {});
        return expect(result).toEqual({
          nodes: output,
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false,
          },
          totalCountFn: expect.any(Function),
        });
      });
    });

    describe('many limit no paging', () => {
      @Resolver(() => TestResolverDTO)
      class TestResolver extends ReadRelationsResolver(TestResolverDTO, {
        one: { relation: { DTO: TestRelationDTO }, custom: { DTO: TestRelationDTO, relationName: 'other' } },
        many: {
          relations: { DTO: TestRelationDTO, pagingStrategy: PagingStrategies.NONE },
          customs: { DTO: TestRelationDTO, pagingStrategy: PagingStrategies.NONE, relationName: 'others' },
        },
      }) {
        constructor(service: TestService) {
          super(service);
        }
      }

      it('should call the service queryRelations with the provided dto', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: NonePagingQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(TestRelationDTO, 'relations', deepEqual([dto]), objectContaining({ ...query })),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryRelations(dto, query, {});
        return expect(result).toEqual(output);
      });

      it('should call the service findRelation with the provided dto and correct relation name', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: OffsetQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(TestRelationDTO, 'others', deepEqual([dto]), objectContaining(query)),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryCustoms(dto, query, {});
        return expect(result).toEqual(output);
      });
    });
  });
});
