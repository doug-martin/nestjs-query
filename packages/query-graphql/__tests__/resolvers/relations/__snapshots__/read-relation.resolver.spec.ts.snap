// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadRelationsResolver many should not add read methods if disableRead is true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver many should set the field to nullable if set to true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  relations(
    """Limit or page results."""
    paging: CursorPaging = {first: 10}

    """Specify to filter the records returned."""
    filter: TestRelationDTOFilter = {}

    """Specify to sort results."""
    sorting: [TestRelationDTOSort!] = []
  ): TestResolverDTORelationsConnection
}

input CursorPaging {
  """Paginate before opaque cursor"""
  before: ConnectionCursor

  """Paginate after opaque cursor"""
  after: ConnectionCursor

  """Paginate first"""
  first: Int

  """Paginate last"""
  last: Int
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

input TestRelationDTOFilter {
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input TestRelationDTOSort {
  field: TestRelationDTOSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TestRelationDTOSortFields {
  id
  testResolverId
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type TestResolverDTORelationsConnection {
  """Paging information"""
  pageInfo: PageInfo!

  """Array of edges."""
  edges: [TestRelationDTOEdge!]!
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver many should use an offset connection if pagingStrategy is offset 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  relations(
    """Limit or page results."""
    paging: OffsetPaging = {limit: 10}

    """Specify to filter the records returned."""
    filter: TestRelationDTOFilter = {}

    """Specify to sort results."""
    sorting: [TestRelationDTOSort!] = []
  ): TestResolverDTORelationsConnection
}

input OffsetPaging {
  """Limit the number of records returned"""
  limit: Int

  """Offset to start returning records from"""
  offset: Int
}

input TestRelationDTOFilter {
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input TestRelationDTOSort {
  field: TestRelationDTOSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TestRelationDTOSortFields {
  id
  testResolverId
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type TestResolverDTORelationsConnection {
  """Paging information"""
  pageInfo: OffsetPageInfo!

  """Array of nodes."""
  nodes: [TestRelationDTO!]!
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver many should use the dtoName if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  tests(
    """Limit or page results."""
    paging: CursorPaging = {first: 10}

    """Specify to filter the records returned."""
    filter: TestRelationDTOFilter = {}

    """Specify to sort results."""
    sorting: [TestRelationDTOSort!] = []
  ): TestResolverDTOTestsConnection!
}

input CursorPaging {
  """Paginate before opaque cursor"""
  before: ConnectionCursor

  """Paginate after opaque cursor"""
  after: ConnectionCursor

  """Paginate first"""
  first: Int

  """Paginate last"""
  last: Int
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

input TestRelationDTOFilter {
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input TestRelationDTOSort {
  field: TestRelationDTOSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TestRelationDTOSortFields {
  id
  testResolverId
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type TestResolverDTOTestsConnection {
  """Paging information"""
  pageInfo: PageInfo!

  """Array of edges."""
  edges: [TestRelationDTOEdge!]!
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver many should use the object type name 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  relations(
    """Limit or page results."""
    paging: CursorPaging = {first: 10}

    """Specify to filter the records returned."""
    filter: TestRelationDTOFilter = {}

    """Specify to sort results."""
    sorting: [TestRelationDTOSort!] = []
  ): TestResolverDTORelationsConnection!
}

input CursorPaging {
  """Paginate before opaque cursor"""
  before: ConnectionCursor

  """Paginate after opaque cursor"""
  after: ConnectionCursor

  """Paginate first"""
  first: Int

  """Paginate last"""
  last: Int
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

input TestRelationDTOFilter {
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input TestRelationDTOSort {
  field: TestRelationDTOSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TestRelationDTOSortFields {
  id
  testResolverId
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type TestResolverDTORelationsConnection {
  """Paging information"""
  pageInfo: PageInfo!

  """Array of edges."""
  edges: [TestRelationDTOEdge!]!
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver one should not add read one methods if disableRead is true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver one should set the field to nullable if set to true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  relation: TestRelationDTO
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver one should use the dtoName if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  test: TestRelationDTO!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver one should use the object type name 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  relation: TestRelationDTO!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReadRelationsResolver should not add read methods if one and many are empty 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;
